### PR TITLE
Ilan/add permissions contract

### PR DESCRIPTION
**Pull request by ilanDoron**
added mock permissions contract in mockcontracts folder.
added tests for all functions.

### CODE REVIEW

File: contracts/Permissions.sol [hunk 1]
```diff
@@ -0,0 +1,102 @@
+pragma solidity ^0.4.0;
+
+
+contract Permissions {
+
+    address public Admin;
```
Comment (yaronvel) [inline contracts/Permissions.sol]: variable name should start with lower case letter

File: contracts/Permissions.sol [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    address public proposedAdmin;
+    mapping(address=>bool) public Operators;
```
Comment (yaronvel) [inline contracts/Permissions.sol]: also here and everywhere

File: contracts/Permissions.sol [hunk 3]
```diff
@@ -0,0 +1,102 @@
+pragma solidity ^0.4.0;
+
+
+contract Permissions {
```
Comment (yaronvel) [inline contracts/Permissions.sol]: think of better name. maybe PermissionLevels? think of something better.

File: contracts/Permissions.sol [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    function requestOwnershipTransfer( address newAdmin ) public
+        onlyAdmin
+    {
+        proposedAdmin = newAdmin;
+    }
+
+    function acceptOwnerShip() public {
+        if (msg.sender != proposedAdmin) revert ();
```
Comment (yaronvel) [inline contracts/Permissions.sol]: change to `require`

File: contracts/Permissions.sol [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    modifier onlyAlerter() {
+        require (Alerters[msg.sender]);
+        _;
+    }
+
+    function requestOwnershipTransfer( address newAdmin ) public
+        onlyAdmin
+    {
+        proposedAdmin = newAdmin;
```
Comment (yaronvel) [inline contracts/Permissions.sol]: should be logged as an event

File: contracts/Permissions.sol [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        require (Operators[msg.sender]);
+        _;
+    }
+
+    modifier onlyAlerter() {
+        require (Alerters[msg.sender]);
+        _;
+    }
+
+    function requestOwnershipTransfer( address newAdmin ) public
```
Comment (yaronvel) [inline contracts/Permissions.sol]: it is not really request. think of better name. Maybe see what bancor named it.

File: contracts/Permissions.sol [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        onlyAdmin
+    {
+        proposedAdmin = newAdmin;
+    }
+
+    function acceptOwnerShip() public {
```
Comment (yaronvel) [inline contracts/Permissions.sol]: ownership is one word

File: contracts/Permissions.sol [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        if (msg.sender != proposedAdmin) revert ();
+        Admin = proposedAdmin;
```
Comment (yaronvel) [inline contracts/Permissions.sol]: log as event

File: contracts/Permissions.sol [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    function addAlerter( address newAlerter ) public
+        onlyAdmin
+    {
+        Alerters[newAlerter] = true;
```
Comment (yaronvel) [inline contracts/Permissions.sol]: should log as event

File: contracts/Permissions.sol [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        AlertersGroup.push(newAlerter);
```
Comment (yaronvel) [inline contracts/Permissions.sol]: should check that array size < 100. if not throw.

File: contracts/Permissions.sol [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    function removeAlerter ( address toRemove ) public
```
Comment (yaronvel) [inline contracts/Permissions.sol]: `toRemove` is not a good name. Maybe just call it `address alerter`.

File: contracts/Permissions.sol [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        //TODO - return success?
+        //TODO - can alerter remove himself? or only admin?
+        if (!Alerters[toRemove]) revert();
+
+        for (uint i = 0; i < AlertersGroup.length; ++i)
+        {
+            if (AlertersGroup[i] == toRemove)
+            {
+                AlertersGroup[i] = AlertersGroup[AlertersGroup.length - 1];
+                AlertersGroup.length -= 1;
```
Comment (yaronvel) [inline contracts/Permissions.sol]: there is `--` opertor

File: contracts/Permissions.sol [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        Alerters[newAlerter] = true;
+        AlertersGroup.push(newAlerter);
+    }
+
+    function removeAlerter ( address toRemove ) public
+        onlyAdmin
+    {
+        //TODO - return success?
+        //TODO - can alerter remove himself? or only admin?
+        if (!Alerters[toRemove]) revert();
```
Comment (yaronvel) [inline contracts/Permissions.sol]: log as event

File: contracts/Permissions.sol [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+            if (AlertersGroup[i] == toRemove)
+            {
+                AlertersGroup[i] = AlertersGroup[AlertersGroup.length - 1];
+                AlertersGroup.length -= 1;
+                break;
+            }
+        }
+    }
+
+    function addOperator( address newOperator ) public
```
Comment (yaronvel) [inline contracts/Permissions.sol]: same comments as in alerter

File: contracts/Permissions.sol [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+            if (OperatorsGroup[i] == toRemove)
+            {
+                OperatorsGroup[i] = OperatorsGroup[OperatorsGroup.length - 1];
+                OperatorsGroup.length -= 1;
+                break;
+            }
+        }
+    }
+
+    function getAlerters () public view returns (address[] alerters) {
```
Comment (yaronvel) [inline contracts/Permissions.sol]: AlertersGroup is public variable. But I am not sure what it returns. Whether it is the entire array, or just a function from index to value. Please find out. If it returns the entire array, no need for these functions.

File: contracts/Permissions.sol (hunk 12 above)
Comment (ilanDoron) [reply contracts/Permissions.sol]: don't understand this remark

Comment (yaronvel) [reply contracts/Permissions.sol]: Do 'x--' and not 'x-=1'

**Pull request merged**